pkg/actuators/machine/instances_test.go

```diff
@@ -58,35 +59,128 @@ func TestRemoveDuplicatedTags(t *testing.T) {
 			},
 		},
 		{
-			tagList: []awsproviderv1.TagSpecification{
+			name:            "with nil infra should  return default tags",
+			machineSpecTags: []awsproviderv1.TagSpecification{},
+			infra:           nil, // should work with nil infra object
+			expected: []*ec2.Tag{
+				{Key: aws.String("kubernetes.io/cluster/clusterID"), Value: aws.String("owned")},
+				{Key: aws.String("Name"), Value: aws.String("machineName")},
+			},
+		},
+		{
+			name: "should filter out bad tags from provider spec",
+			machineSpecTags: []awsproviderv1.TagSpecification{
 				{Name: "Name", Value: "badname"},
 				{Name: "kubernetes.io/cluster/badid", Value: "badvalue"},
 				{Name: "good", Value: "goodvalue"},
 			},
+			infra: nil,
+			// Invalid tags get dropped and the valid clusterID and Name get applied last.
+			expected: []*ec2.Tag{
+				{Key: aws.String("good"), Value: aws.String("goodvalue")},
+				{Key: aws.String("kubernetes.io/cluster/clusterID"), Value: aws.String("owned")},
+				{Key: aws.String("Name"), Value: aws.String("machineName")},
+			},
+		},
+		{
+			name:            "should filter out bad tags from infra object",
+			machineSpecTags: []awsproviderv1.TagSpecification{},
+			infra: &configv1.Infrastructure{
+				Status: configv1.InfrastructureStatus{
+					PlatformStatus: &configv1.PlatformStatus{
+						AWS: &configv1.AWSPlatformStatus{
+							ResourceTags: []configv1.AWSResourceTag{
+								{
+									Key:   "kubernetes.io/cluster/badid",
+									Value: "badvalue",
+								},
+								{
+									Key:   "Name",
+									Value: "badname",
+								},
+								{
+									Key:   "good",
+									Value: "goodvalue",
+								},
+							},
+						},
+					},
+				},
+			},
 			// Invalid tags get dropped and the valid clusterID and Name get applied last.
 			expected: []*ec2.Tag{
 				{Key: aws.String("good"), Value: aws.String("goodvalue")},
 				{Key: aws.String("kubernetes.io/cluster/clusterID"), Value: aws.String("owned")},
 				{Key: aws.String("Name"), Value: aws.String("machineName")},
 			},
 		},
+		{
+			name: "tags from infrastructure object should have precedence",
+			machineSpecTags: []awsproviderv1.TagSpecification{
+				{Name: "Name", Value: "badname"},
+				{Name: "kubernetes.io/cluster/badid", Value: "badvalue"},
+				{Name: "good", Value: "should-be-overwritten"},
+			},
+			infra: &configv1.Infrastructure{
+				Status: configv1.InfrastructureStatus{
+					PlatformStatus: &configv1.PlatformStatus{
+						AWS: &configv1.AWSPlatformStatus{
+							ResourceTags: []configv1.AWSResourceTag{
+								{
+									Key:   "good",
+									Value: "goodvalue",
+								},
+							},
+						},
+					},
+				},
+			},
+			expected: []*ec2.Tag{
+				{Key: aws.String("good"), Value: aws.String("goodvalue")},
+				{Key: aws.String("kubernetes.io/cluster/clusterID"), Value: aws.String("owned")},
+				{Key: aws.String("Name"), Value: aws.String("machineName")},
+			},
+		},
```

Comment:
Ok so we need to reverse this test based on the latest version of the enhancement that was agreed

Reply:
thanks, I'll fix it in a minute